library/kotlin/src/io/envoyproxy/envoymobile/filters/FilterDataStatus.kt

```diff
@@ -45,4 +45,14 @@ sealed class FilterDataStatus {
    * continuing processing, and will handle their own buffering.
    */
   class StopIterationNoBuffer : FilterDataStatus()
+
+  /**
+   * Resume previously-stopped iteration, possibly forwarding headers, if iteration was previously
+   * stopped during an on*Headers invocation.
+   *
+   * It is an error to return ResumeIteration if iteration is not currently stopped, and it is
+   * an error to include headers if headers have already been forwarded to the next filter
+   * (i.e. iteration was stopped during an on*Data invocation instead of on*Headers).
+   */
+  class ResumeIteration<T : Headers>(val headers: T?, val data: ByteBuffer) : FilterDataStatus()
```

Comment:
If the caller omits `headers` here (passes `nil`), that will result in the underlying filter manager passing its internally buffered headers, correct? If so, I think that's worth documenting above

Reply:
Another option is for that to be an error (and I can update the documentation accordingly). To simplify reasoning, the contract could _always_ be that all forwarded entities are passed back as part of a return from a filter invocation. If it's not explicitly passed back, it hasn't been forwarded.